File: problems/0001-two-sum/solution.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        Problem: 0001 - Two Sum

        Given an array of integers and a target value, return the indices of 
        the two numbers such that they add up to the target.

        Approach:
        - Use a hash map (dictionary) to store previously seen numbers and their indices.
        - For each number in the array:
            - Calculate the complement (target - current number)
            - If the complement is already in the map, return the pair of indices.
            - Otherwise, store the current number with its index in the map.

        Time Complexity: O(n)
        Space Complexity: O(n)
        """

        map_ = {}  # Dictionary to store seen numbers and their indices.
        for idx in range(len(nums)):
            remain = target - nums[idx] # The number we are looking to find.
            if remain in map_:
		# Found the complement -- return indices.
                return [map_[remain], idx]
	    # Store the current number with its index.
            map_[nums[idx]] = idx

File: problems/0001-two-sum/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        /*
        Problem: 0001 - Two Sum.

        Given an array of integers and a target value, return the indices of
        the two numbers such that they add up to the target.

        Approach:
        - Use an unordered_map (hash map) to track values and their indices.
        - For each element, calculate its complement (target - num).
        - If the complement is found in the map, return the two indices.
        - Otherwise, store the current value and its index in the map.

        Time Complexity: O(n).
        Space Complexity: O(n).
        */

        unordered_map<int, int> map; // Value → index.

        for (int i = 0; i < nums.size(); ++i) {
            int remain = target - nums[i]; // Compute complement.
            if (map.find(remain) != map.end()) {
                // Found the complement.
                return {map[remain], i};
            }
            // Store current number and index.
            map[nums[i]] = i;
        }

        // Should not reach here since one valid answer always exists.
        return {};
    }
};
